flashlight/fl/tensor/backend/jit/JitTensorBase.cpp

```diff
@@ -7,41 +7,134 @@
 
 #include "flashlight/fl/tensor/backend/jit/JitTensorBase.h"
 
+#include <memory>
 #include <sstream>
+#include <stdexcept>
 
 #define FL_JIT_TENSOR_UNIMPLEMENTED \
   throw std::invalid_argument(      \
       "JitTensorBase::" + std::string(__func__) + " - unimplemented.");
 
 namespace fl {
 
-struct JitTensorBase::SharedData {
+// represents the data referred to by indexinges
+struct DataStorage {
   Node* node;
-  SharedData(Node* node) : node(node) {
-    node->incRefCount();
+
+  DataStorage(Node* node) : node(node) {
+    node->incRefCount(); // shallow copies counts as 1 use
   }
-  ~SharedData() {
+
+  ~DataStorage() {
+    node->decRefCount();
+  }
+
+  void replaceNode(Node* newNode) {
+    newNode->incRefCount();
     node->decRefCount();
+    node = newNode;
+  }
+};
+
+// data shared among shallow copies and views of a tensor.
+class JitTensorBase::SharedData {
+  // 1. Why `shared_ptr<DataStorage> dataStorage_`?
+  //   Indexing essentially creates a shallow-copy that views the indexed region
+  //   of the original tensor. So any update to the original tensor should
+  //   reflect as an update to dataNode for all shallow copies.
+  // 2. Why `indexings_`?
+  //   DataNode update requires creating a new viewNode_ based on the old
+  //   indexings (recall SSA semantics).
+  // 3. Why `oldDataNode_`?
+  //   When underlying data node changes, not all SharedData objects are
+  //   informed directly, so we lazily perform (2), and use `oldDataNode_` to
+  //   guide us.
+  std::shared_ptr<DataStorage> dataStorage_;
+  // TODO consider
+  // 1. using an immutable linked-list here to speed things up
+  // 2. making `std::vector<Index>` into an immutable class and use it as
+  //    shared_ptr (to avoid copying when passing to ViewNodoe), since it's all
+  //    readonly -- JitTensor gives us free immutability for Tensor index.
+  // 3. index merging, e.g., (1, 3)(1, 2) --> (2, 3). Maybe as an optimization
+  //    pass, need to think more.
+  std::vector<std::vector<Index>> indexings_{};
+  Node* oldDataNode_{nullptr}; // nullptr iff no indexings
```

Comment:
Why use a null pointer for one an an `optional` for the below? Is there a difference between `std::nullopt` and a valid `optional` that has the value `nullptr`? If so, seems a tad error-prone

Reply:
Duplicated comment:)